File: Graveyard/src/de/hannisoft/gaveplan/model/GraveMap.java
package de.hannisoft.gaveplan.model;

import java.util.Calendar;
import java.util.HashSet;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

public class GraveMap {
    private final String field;
    private Set<GraveSite> graveSites = new HashSet<>();

    private int rowCount = -1;
    private int placeCount = -1;
    private int deltaRow = 0;
    private int deltaPlace = 0;
    private boolean finished = false;

    private Grave[][] graveArray = null;

    public GraveMap(String field) {
        this.field = field;
    }

    public void finishEdit(Map<String, PlanElement> fieldElements) {
        if (finished) {
            return;
        }
        fillupMissingGraves();
        initMinMaxValues(fieldElements);
        fillGraveArray();
        setGraveClasses();
        finished = true;
    }

    private void fillupMissingGraves() {
        for (GraveSite grave : graveSites) {
            HashSet<String> places = new HashSet<>();
            for (Grave plc : grave.getGraves()) {
                places.add(plc.getId());
            }
            for (int i = 0; i < grave.getRowSize(); i++) {
                for (int j = 0; j < grave.getPlaceSize(); j++) {
                    try {
                        int row = (grave.getRowInt() + i);
                        String rowStr = String.format("%02d", row);
                        int plc = (grave.getPlaceInt() + j);
                        String placeStr = String.format("%02d", plc);
                        if (!places.contains(rowStr + "/" + placeStr)) {
                            grave.getGraves().add(new Grave(grave, rowStr, placeStr));
                        }
                    } catch (Exception e) {
                        System.err.println(e.getClass().getSimpleName() + " while filling missing places of " + grave.toString()
                                + " into GraveMap: " + e.getMessage());
                        e.printStackTrace();
                    }
                }
            }

        }
    }

    private void initMinMaxValues(Map<String, PlanElement> fieldElements) {
        PlanElement fe = fieldElements.get(field);
        if (fe == null) {
            initMinMaxValuesFromGaves();
        } else {
            System.out.println("Found FieldElement of GraveMap " + field + " " + fe.toString());
            rowCount = fe.getMaxRow() - fe.getMinRow() + (fe.getMinRow() < 0 ? 0 : 1);
            placeCount = fe.getMaxPlace() - fe.getMinPlace() + (fe.getMinPlace() < 0 ? 0 : 1);
            deltaRow = fe.getMinRow() < 0 ? Math.abs(fe.getMinRow()) : 0;
            deltaPlace = fe.getMinPlace() < 0 ? Math.abs(fe.getMinPlace()) : 0;
            System.out.println("Initialized MinMaxValues of GraveMap " + field + ": rowCount=" + rowCount + " / placeCount="
                    + placeCount + " / " + deltaRow + " / deltaPlace=" + deltaPlace);// TODO Auto-generated method stub
            // initMinMaxValuesFromGaves();
            System.out.println("");
        }

    }

    private void initMinMaxValuesFromGaves() {
        int minRow = 1;
        int maxRow = 1;
        int minPlace = 1;
        int maxPlace = 1;
        for (GraveSite graveSite : graveSites) {
            for (Grave grave : graveSite.getGraves()) {
                int row = grave.getRowInt();
                if (row < minRow) {
                    minRow = row;
                }
                if (row > maxRow) {
                    maxRow = row;
                }

                int plc = grave.getPlaceInt();
                if (plc < minPlace) {
                    minPlace = plc;
                }
                if (plc > maxPlace) {
                    maxPlace = plc;
                }
            }
        }
        System.out.println("Found MinMaxValues of GraveMap " + field + ": Rows=[" + minRow + ", " + maxRow + "] Grave=["
                + minPlace + ", " + maxPlace + "]");
        rowCount = maxRow - minRow + (minRow < 0 ? 0 : 1);
        placeCount = maxPlace - minPlace + (minPlace < 0 ? 0 : 1);
        deltaRow = minRow < 0 ? Math.abs(minRow) : 0;
        deltaPlace = minPlace < 0 ? Math.abs(minPlace) : 0;
        System.out.println("Initialized MinMaxValues of GraveMap " + field + ": rowCount=" + rowCount + " / placeCount="
                + placeCount + " / " + deltaRow + " / deltaPlace=" + deltaPlace);
    }

    private void fillGraveArray() {
        graveArray = new Grave[rowCount][placeCount];
        for (GraveSite graveSite : graveSites) {
            for (Grave grave : graveSite.getGraves()) {
                try {
                    int row = deltaRow + grave.getRowInt() - (grave.getRowInt() > 0 ? 1 : 0);
                    int plc = deltaPlace + grave.getPlaceInt() - (grave.getPlaceInt() > 0 ? 1 : 0);
                    graveArray[row][plc] = grave;
                } catch (Exception e) {
                    System.err.println(e.getClass().getSimpleName() + " while filling Grave " + grave.toString()
                            + " into GraveMap: " + e.getMessage());
                    e.printStackTrace();
                }
            }
        }
    }

    private void setGraveClasses() {
        for (int i = 0; i < graveArray.length; i++) {
            Grave[] graveRow = graveArray[i];
            for (int j = 0; j < graveRow.length; j++) {
                Grave grave = null;
                try {
                    grave = graveRow[j];
                    if (grave == null) {
                        continue;
                    }
                    GraveSite graveSite = grave.getGraveSite();
                    boolean isBroached = graveSite.isBroached();
                    if (isBroached) {
                        grave.addClasses(GraveClass.BROACHED);
                    } else if (!grave.isEmpty()) {
                        grave.addClasses(GraveClass.BUSY);
                    }

                    if (graveSite.isStele()) {
                        grave.addClasses(GraveClass.STELE);
                    }

                    if (grave.isRef()) {
                        grave.addClasses(GraveClass.REF);
                    }

                    if (!hasNeighborGraveInEqualGraveSite(i, j + 1, graveSite)) {
                        grave.addClasses(GraveClass.W);
                    }

                    if (!hasNeighborGraveInEqualGraveSite(i, j - 1, graveSite)) {
                        grave.addClasses(GraveClass.O);
                    }

                    if (!hasNeighborGraveInEqualGraveSite(i + 1, j, graveSite)) {
                        grave.addClasses(GraveClass.N);
                    }

                    if (!hasNeighborGraveInEqualGraveSite(i - 1, j, graveSite)) {
                        grave.addClasses(GraveClass.S);
                    }

                    if (!isBroached && grave.getDeceased() == null) {
                        grave.addClasses(GraveClass.FREE);
                    }

                    addRuntime(grave);
                } catch (Exception e) {
                    System.err.println(e.getClass().getSimpleName() + " while initialising Grave's classes of " + grave + ": "
                            + e.getMessage());
                    e.printStackTrace();
                }
            }
        }
    }

    private void addRuntime(Grave grave) {
        if (grave != null && grave.getGraveSite() != null && grave.getGraveSite().getValidFrom() != null) {
            Calendar validTo = Calendar.getInstance(Locale.GERMANY);
            validTo.setTime(grave.getGraveSite().getValidTo());
            Calendar now = Calendar.getInstance(Locale.GERMANY);
            int diff = validTo.get(Calendar.YEAR) - now.get(Calendar.YEAR);
            grave.setRuntimeYear(diff);
        }
    }

    private boolean hasNeighborGraveInEqualGraveSite(int row, int place, GraveSite graveSite) {
        Grave neighbor = null;
        if (row >= 0 && row < rowCount && place >= 0 && place < placeCount) {
            neighbor = graveArray[row][place];
        }

        return neighbor != null && neighbor.getGraveSite() == graveSite;
    }

    public void addGraveSite(GraveSite graveSite) {
        graveSites.add(graveSite);
    }

    public Set<GraveSite> graveSites() {
        return graveSites;
    }

    public String getField() {
        return field;
    }

    public int getRowCount() {
        return rowCount;
    }

    public int getPlaceCount() {
        return placeCount;
    }

    public int getDeltaRow() {
        return deltaRow;
    }

    public int getDeltaPlace() {
        return deltaPlace;
    }

    public Grave getGrave(int row, int place) {
        return graveArray[row][place];
    }
}

File: src/main/kotlin/de/hannisoft/graveplan/model/GraveMap.kt
package de.hannisoft.graveplan.model

import java.util.*
import kotlin.math.abs

class GraveMap(private val fieldName: String) {
    private var field: PlanElement? = null
    private val graveSites: MutableSet<GraveSite> = mutableSetOf()

    private var rowCount: Int = -1
    private var placeCount: Int = -1
    private var deltaRow: Int = 0
    private var deltaPlace: Int = 0
    private var finished: Boolean = false

    private lateinit var graveArray: Array<Array<Grave?>>

    fun finishEdit(fieldElements: Map<String, PlanElement>) {
        if (finished) return

        fillupMissingGraves()
        initMinMaxValues(fieldElements)
        fillGraveArray()
        setGraveClasses()
        finished = true
    }

    private fun fillupMissingGraves() {
        graveSites.forEach { graveSite ->
            val places = graveSite.graves.map { it.id }.toSet()
            for (i in 0 until graveSite.rowSize) {
                for (j in 0 until graveSite.placeSize) {
                    try {
                        val grave = Grave(graveSite, graveSite.rowInt + i, graveSite.placeInt + j)
                        if (!places.contains(graveSite.id)) {
                            graveSite.graves.add(grave)
                        }
                    } catch (e: Exception) {
                        println("${e.javaClass.simpleName} while filling missing places of $graveSite into GraveMap: ${e.message}")
                        e.printStackTrace()
                    }
                }
            }
        }
    }

    private fun initMinMaxValues(fieldElements: Map<String, PlanElement>) {
        field = fieldElements[fieldName]
        if (field == null) {
            initMinMaxValuesFromGraves()
        } else {
            println("Found FieldElement of GraveMap $field")
            rowCount = field!!.maxRow - field!!.minRow + if (field!!.minRow < 0) 0 else 1
            placeCount = field!!.maxPlace - field!!.minPlace + if (field!!.minPlace < 0) 0 else 1
            deltaRow = if (field!!.minRow < 0) abs(field!!.minRow) else 0
            deltaPlace = if (field!!.minPlace < 0) abs(field!!.minPlace) else 0
            println("Initialized MinMaxValues of GraveMap $field: rowCount=$rowCount / placeCount=$placeCount / $deltaRow / deltaPlace=$deltaPlace")
        }
    }

    private fun initMinMaxValuesFromGraves() {
        var minRow = 1
        var maxRow = 1
        var minPlace = 1
        var maxPlace = 1

        for (graveSite in graveSites) {
            for (grave in graveSite.graves) {
                val row = grave.row
                minRow = minOf(minRow, row)
                maxRow = maxOf(maxRow, row)

                val plc = grave.place
                minPlace = minOf(minPlace, plc)
                maxPlace = maxOf(maxPlace, plc)
            }
        }
        println("Found MinMaxValues of GraveMap $field: Rows=[$minRow, $maxRow] Grave=[$minPlace, $maxPlace]")

        rowCount = maxRow - minRow + if (minRow < 0) 0 else 1
        placeCount = maxPlace - minPlace + if (minPlace < 0) 0 else 1
        deltaRow = if (minRow < 0) abs(minRow) else 0
        deltaPlace = if (minPlace < 0) abs(minPlace) else 0

        field = PlanElement(this.hashCode(), "feld", minRow, maxRow, minPlace, maxPlace, fieldName)
        println("Initialized MinMaxValues of GraveMap $field: rowCount=$rowCount / placeCount=$placeCount / $deltaRow / deltaPlace=$deltaPlace")
    }

    private fun fillGraveArray() {
        graveArray = Array(rowCount) { arrayOfNulls<Grave>(placeCount) }
        for (graveSite in graveSites) {
            for (grave in graveSite.graves) {
                try {
                    val row = deltaRow + grave.row - if (grave.row > 0) 1 else 0
                    val plc = deltaPlace + grave.place - if (grave.place > 0) 1 else 0
                    graveArray[row][plc] = grave
                } catch (e: Exception) {
                    println("${e.javaClass.simpleName} while filling Grave $grave into GraveMap: ${e.message}")
                    e.printStackTrace()
                }
            }
        }
    }

    private fun setGraveClasses() {
        graveArray.forEachIndexed { i, graveRow ->
            graveRow.forEachIndexed { j, grave ->
                try {
                    if (grave == null)
                        return@forEachIndexed

                    val graveSite = grave.graveSite

                    val isBroached = graveSite.isBroached()
                    if (isBroached) {
                        grave.addClass(GraveClass.BROACHED)
                    } else if (!grave.isEmpty()) {
                        grave.addClass(GraveClass.BUSY)
                    }

                    if (graveSite.isStele()) {
                        grave.addClass(GraveClass.STELE)
                    }

                    if (grave.isRef()) {
                        grave.addClass(GraveClass.REF)
                    }

                    if (!hasNeighborGraveInEqualGraveSite(i, j + 1, graveSite)) {
                        grave.addClass(GraveClass.W)
                    }
                    if (!hasNeighborGraveInEqualGraveSite(i, j - 1, graveSite)) {
                        grave.addClass(GraveClass.O)
                    }
                    if (!hasNeighborGraveInEqualGraveSite(i + 1, j, graveSite)) {
                        grave.addClass(GraveClass.N)
                    }
                    if (!hasNeighborGraveInEqualGraveSite(i - 1, j, graveSite)) {
                        grave.addClass(GraveClass.S)
                    }

                    if (!isBroached && grave.deceased == null) {
                        grave.addClass(GraveClass.FREE)
                    }

                    addRuntime(grave)
                } catch (e: Exception) {
                    println("${e.javaClass.simpleName} while initialising Grave's classes of $grave: ${e.message}")
                    e.printStackTrace()
                }
            }
        }
    }

    private fun addRuntime(grave: Grave?) {
        if (grave?.graveSite?.validFrom != null) {
            val validTo = Calendar.getInstance(Locale.GERMANY).apply {
                time = grave.graveSite.validTo
            }
            val now = Calendar.getInstance(Locale.GERMANY)
            val diff = validTo[Calendar.YEAR] - now[Calendar.YEAR]
            grave.runtimeYear = diff
        }
    }

    private fun hasNeighborGraveInEqualGraveSite(row: Int, place: Int, graveSite: GraveSite): Boolean {
        val neighbor =
            if (row in 0 until rowCount && place in 0 until placeCount) {
                graveArray[row][place]
            } else
                null
        return neighbor != null && neighbor.graveSite == graveSite
    }

    fun addGraveSite(graveSite: GraveSite) {
        graveSites.add(graveSite)
    }

    fun graveSites(): Set<GraveSite> = graveSites

    fun getFieldName(): String = field?.name ?: fieldName

    fun getField(): PlanElement? = field

    fun getGrave(row: Int, place: Int): Grave? = graveArray[row][place]
}